---
title: "AP/NTS Gfral TRAP-seq"
author: "Alan Rupp"
date: "`r strftime(Sys.time(), '%B %d, %Y')`"
output: pdf_document
geometry: margin=0.5in
---

Analyzing TRAP-seq data from *Gfral<sup>Cre</sup>*::L10 mice dissected from the AP/NTS. Data is combined from 2 runs, one of which had mice that were treated with either GDF15 or Vehicle. I'm looking for genes enriched in *Gfral* neurons and genes that are regulated in *Gfral* neurons in response to GDF15 treatment.

```{r setup, include = FALSE}
knitr::opts_chunk$set(warning = FALSE, echo = FALSE, message = FALSE)
library(knitr); library(kableExtra)
library(tidyverse)
library(ggrepel)
map(list.files("tools/rnaseq/R", full.names = TRUE), source)
```

## Samples
```{r read-info}
# read in info and metadata
files <- list.files(pattern = "*ReadsPerGene.out.tab",
                    recursive = TRUE,
                    full.names = TRUE)
samples <- str_extract(files, "Sample_[0-9]+")

run_files <- list.files(pattern = "Run.+csv",
                       recursive = TRUE,
                       full.names = TRUE)
info <- map(run_files, ~ read_csv(.x, skip = 18)) %>%
    bind_rows() %>%
    filter(!duplicated(Sample_ID)) %>%
    filter(Sample_ID %in% samples) %>%
    select(Sample_ID, Description) %>%
    mutate(cells = ifelse(grepl("B", Description), "Bead", "Sup")) %>%
    mutate(cells = factor(cells, levels = c("Sup", "Bead"))) %>%
    mutate(run = c(rep("Run_1905", 2), rep("Run_2487", 10))) %>%
    mutate(pair = paste0(run, str_extract(Description, "[d,e]"),
                         str_extract(Description, "[0-9]$"))) %>%
    mutate(treatment = ifelse(grepl("Gdf15", Description),
                              "GDF15", "Vehicle")) %>%
    mutate(treatment = factor(treatment, levels = c("Vehicle", "GDF15")))

info %>%
    select(-Description, -pair) %>%
    kable() %>%
    kable_styling(., position = "left", full_width = FALSE)
```

```{r read-data}
# - bring in data -------------------------------------------------------------
genes <- read_csv("data/genes.csv")
counts <- read_counts(files)
```

\pagebreak

# Quality control
## Positive controls
```{r positive-controls, fig.width = 3, fig.height = 3}
# - controls -------------------------------------------------------------------
positive_control_genes <- c("Cre", "Gfp_L10a", "Gfral")
positive_controls <- filter(genes, gene_name %in% positive_control_genes)$gene_id

make_cpm(counts) %>%
  filter(gene_id %in% positive_controls) %>%
  left_join(genes, by = "gene_id") %>%
  gather(starts_with("Sample"), key = "sample", value = "counts") %>%
  left_join(., select(info, Sample_ID, cells, pair, treatment),
            by = c("sample" = "Sample_ID")) %>%
  mutate(gene_name = factor(gene_name,
                            levels = c("Gfral", "Cre", "Gfp_L10a"),
                            labels = c("Gfral", "Cre", "EGFP:L10a"))) %>%
  ggplot(aes(x = cells, y = counts, group = pair)) +
  geom_point() +
  geom_line() +
  scale_y_continuous(trans = "log2",
                     breaks = 2^c(-3, 0, 3, 6),
                     labels = c("0.125", "1", "8", "64")) +
  labs(y = expression("Expression (CPM)"), x = element_blank()) +
  theme_classic() +
  theme(legend.position = c(0.15, 0.85),
        axis.text.x = element_text(color = "black", size = 10, angle = 45,
                                   hjust = 1, vjust = 1),
        axis.text.y = element_text(color = "black"),
        strip.background = element_blank(),
        strip.text = element_text(size = 10)) +
  facet_wrap(~gene_name)

ggsave("figures/panels/line.png", width = 3, height = 3, units = "in", dpi = 600)
```

## Heatmap clustering
```{r heatmap}
heatmap <- function(counts, metadata, annotate, threshold = 1) {
  log2cpm <- make_cpm(counts, log2 = TRUE)
  expressed <- log2cpm %>%
    mutate_if(is.numeric, ~.x >= threshold) %>%
    as.data.frame() %>%
    tibble::column_to_rownames("gene_id")
  expressed <- rowSums(expressed) > 3
  expressed_genes <- names(expressed)[expressed]

  # run clustering
  df <- log2cpm %>%
    as.data.frame() %>%
    tibble::column_to_rownames("gene_id") %>%
    filter(expressed)
  dists <- dist(t(df))
  clusters <- hclust(dists)
  groups <- cutree(clusters, 2)
  metadata <- mutate(metadata, "cut" = groups)

  # set up annotation
  annotation <- select(info, annotate) %>% as.data.frame()
  rownames(annotation) <- info$Sample_ID

  plt <- pheatmap::pheatmap(df, show_rownames = FALSE,
                     annotation_col = annotation,
                     annotation_colors = list(
                       cells = c(Bead = "seagreen", Sup = "gray70"),
                       treatment = c(Vehicle = "gray", GDF15 = "black")
                       )
                      )
  return(list("plot" = plt, "group" = groups))
}

heatmap_cluster <- heatmap(counts, info, c("cells", "treatment"))
heatmap_cluster$plot
```

Samples Sample_85276 and Sample_85277 appear to be outliers.

## tSNE Clustering
```{r tsne, fig.width = 3, fig.height = 3.3}
reduction <- function(counts) {
  df <- counts %>%
    filter(rowSums(counts[,2:ncol(counts)]) > 0) %>%
    mutate_if(is.numeric, ~ log2(.x / sum(.x) * 10^6))
  # remove genes with no expression
  threshold <- rowSums(df[,2:ncol(df)]) > 0
  df <- filter(df, threshold)
  # remove infinites
  df <- filter(df, rowSums(is.infinite(as.matrix(df[,2:ncol(df)]))) == 0)
  # make df with rownames
  df <- df %>% as.data.frame() %>% column_to_rownames("gene_id")

  # run PCA
  choose_pcs <- function(pca, matrix, reps = 100, sig_level = 0.05) {
    mtrx <- as.matrix(matrix)
    randomize <- function(mtrx) {
      result <- sapply(seq(nrow(mtrx)), function(x) sample(mtrx[x, ]))
      return(result)
    }
    find_eigenvalues <- function(matrix) {
      pca <- prcomp(randomize(matrix), scale. = TRUE)
      return(pca$sdev^2)
    }
    eigenvalues <- replicate(reps, find_eigenvalues(mtrx))
    # find the frequency that real eigenvalue is above bootstrap eigenvalue
    prob <- 1 - rowSums(pca$sdev^2 > eigenvalues) / reps
    sig_pcs <- which(prob < sig_level)
    return(sig_pcs)
  }
  print("Running PCA ...")
  pca <- prcomp(t(df), scale. = TRUE)
  print("Choosing PCs by statistical significance ...")
  pcs <- choose_pcs(pca, df)

  # run TSNE from PCA
  set.seed(32)
  tsne <- Rtsne::Rtsne(pca$x[, pcs], perplexity = (nrow(pca$x) - 1) / 3)

  # classify using k-means to find outliers
  print(paste("Running tSNE with perplexity =", round((nrow(pca$x) - 1) / 3, 1),
              "..."))
  tsne <- as.data.frame(tsne$Y) %>%
    set_names("TSNE1", "TSNE2") %>%
    mutate("Sample_ID" = rownames(pca$x))
  groups <- kmeans(select(tsne, starts_with("TSNE")), 2)
  names(groups$cluster) <- rownames(pca$x)

  # plot
  plt <- left_join(tsne, info, by = "Sample_ID") %>%
    ggplot(aes(x = TSNE1, y = TSNE2, color = cells, shape = treatment)) +
    geom_point() +
    geom_text_repel(aes(label = Sample_ID), show.legend = FALSE) +
    scale_color_manual(values = c("gray30", "darkgreen"), name = NULL) +
    scale_shape(name = NULL) +
    theme_void() +
    theme(legend.position = "top",
          legend.direction = "horizontal")
  return(list("plot" = plt, "class" = groups$cluster))
}

tsne <- reduction(counts)
tsne$plot

```

**Sample_85277** appears to be an outlier using tSNE clustering as well. I'm going to remove it and its pair from all analysis.

```{r outliers}
outliers <- c("Sample_85276", "Sample_85277")
counts <- select(counts, -all_of(outliers))
info <- filter(info, !Sample_ID %in% outliers)
```

\pagebreak

# Differential gene expression
I'll use DESeq2 `r packageVersion("DESeq2")` to find differentially expressed genes.

## Enrichment
```{r enrichment}
mtx <- column_to_rownames(as.data.frame(counts), "gene_id")

# finding enriched genes across all treatments
print("Creating DESeq2 object ...")
dds <- DESeq2::DESeqDataSetFromMatrix(mtx, info, ~ pair + cells)
print("Running DESeq ...")
dds <- DESeq2::DESeq(dds)
```

```{r enriched-genes}
print("Finding enriched genes ...")
enrichment <- dds %>%
  DESeq2::results(name = "cells_Bead_vs_Sup") %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  arrange(desc(padj < 0.05), desc(log2FoldChange))

enriched <- filter(enrichment, padj < 0.05 & log2FoldChange > 0)$gene_id
```

```{r cpm}
print("Calculating CPM for each sample ...")
cpm <- dds %>%
  DESeq2::fpm() %>%
  as.data.frame() %>%
  rownames_to_column("gene_id")
```

```{r volcano, fig.width = 3.15, fig.height = 2.5}
df <- left_join(enrichment, genes, by = "gene_id")
df <- mutate(df, gene_name = recode(gene_name, `Gfp_L10a` = "EGFP:L10a"))
color_palette <- wesanderson::wes_palette("Zissou1")

ggplot(df, aes(x = log2FoldChange, y = -log10(padj), color = padj < 0.05)) +
  geom_hline(aes(yintercept = -log10(0.05)), linetype = "dashed") +
  geom_vline(aes(xintercept = 0)) +
  geom_point(stroke = 0, alpha = 0.4, show.legend = FALSE, size = 2) +
  scale_color_manual(values = c(color_palette[4], color_palette[2])) +
  scale_x_continuous(limits = c(-max(abs(enrichment$log2FoldChange),
                                     na.rm = TRUE),
                                max(abs(enrichment$log2FoldChange),
                                    na.rm = TRUE)),
                     expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0),
                     limits = c(0, max(-log10(enrichment$padj),
                                       na.rm = TRUE)*1.1),
                     breaks = seq(from = 0, to = 160, by = 160 / 4)
                     ) +
  annotate("segment", x = 0, xend = 1, y = 160, yend = 160,
           arrow = arrow(length = unit(0.05, "in"))) +
  annotate("segment", x = 0, xend = -1, y = 160, yend = 160,
           arrow = arrow(length = unit(0.05, "in"))) +
  annotate("text", x = 0.2, y = 160,
           label = paste(sum(enrichment$padj < 0.05 & enrichment$log2FoldChange > 0,
                             na.rm = TRUE), "\nenriched"),
           size = 2, hjust = 0) +
  annotate("text", x = -0.2, y = 160,
           label = paste(sum(enrichment$padj < 0.05 & enrichment$log2FoldChange < 0,
                             na.rm = TRUE), "\nde-enriched"),
           size = 2, hjust = 1) +
  geom_text_repel(data = filter(df, gene_id %in% positive_controls),
                  aes(label = gene_name), size = 3, show.legend = FALSE,
                  xlim = c(2.8, NA), color = "black") +
  theme_minimal() +
  theme(panel.grid = element_blank(),
        axis.text = element_text(color = "black")) +
  xlab(expression("Enrichment (log"[2]*" Bead / Sup)")) +
  ylab(expression(italic("P")*" value (-log"[10]*")"))

ggsave("figures/panels/volcano.png", width = 3.15, height = 2.5,
       units = "in", dpi = 600)
```

```{r ma-plot, fig.width = 3.15, fig.height = 2.5}
mean_cpm <- cpm_means(cpm, info, "cells")
mean_cpm <- left_join(mean_cpm, genes, by = "gene_id")
df <- left_join(mean_cpm, enrichment, by = "gene_id")
df <- filter(df, !is.na(padj))
df <- mutate(df, gene_name = recode(gene_name, `Gfp_L10a` = "EGFP:L10a"))

ggplot(df, aes(x = Bead, y = log2FoldChange, color = padj < 0.05)) +
  geom_hline(aes(yintercept = log2(1.5)), linetype = "dashed") +
  geom_point(alpha = 0.4, stroke = 0, size = 2, show.legend = FALSE) +
  scale_x_continuous(trans = "log2", limits = c(1, NA), expand = c(0, 0)) +
  scale_y_continuous(limits = c(0, NA), expand = c(0, 0)) +
  scale_color_manual(values = c(color_palette[4], color_palette[2])) +
  theme_minimal() +
  theme(panel.grid = element_blank(),
        legend.position = c(0.85, 0.75),
        legend.background = element_rect(color = "gray90"),
        axis.text = element_text(color = "black"),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 7)) +
  xlab(expression("Bead expression (CPM)")) +
  ylab(expression("Enrichment (log"[2]*" Bead / Sup)")) +
  coord_flip() +
  annotate("rect", xmin = 1500, xmax = 3500, ymin = 2.75, ymax = 3.25,
           fill = alpha("gray90", 0.5), color = "gray90") +
  annotate("text", y = 3, x = 3000, label = "P ≥ 0.05",
           color = color_palette[4], size = 2) +
  annotate("text", y = 3, x = 2000, label = "P < 0.05",
           color = color_palette[2], size = 2)

ggsave("figures/panels/ma_plot.png", width = 3.15, height = 2.5,
       units = "in", dpi = 600)
```

`r length(enriched)` genes significantly enriched. `r nrow(filter(enrichment, log2FoldChange > log2(1.5) & padj < 0.05))` genes significantly enriched > 1.5-fold change.

```{r enrichment-table}
enrichment %>%
  filter(padj < 0.05 & log2FoldChange > 0) %>%
  left_join(genes, by = "gene_id") %>%
  left_join(mean_cpm, by = c("gene_name", "gene_id", "gene_biotype")) %>%
  filter(gene_biotype == "protein_coding") %>%
  filter(!gene_name %in% positive_control_genes) %>%
  arrange(desc(log2FoldChange)) %>%
  select(gene_name, gene_id, Bead, log2FoldChange, padj) %>%
  slice(1:10) %>%
  rename("CPM" = Bead, "Enrichment" = log2FoldChange, "P" = padj) %>%
  mutate(Enrichment = round(Enrichment, 2), CPM = round(CPM, 1),
         P = format(P, scientific = TRUE, digits = 2)) %>%
  kable(caption = "Top 10 enriched protein-coding genes (excluding Gfral, Cre, and GFP)") %>%
  kable_styling(latex_options = "hold_position")

```

## Enrichment of known neuronal signaling genes
```{r signaling-genes}
anno <- read_csv("tools/rnaseq/data/annotation.csv")
neuropeptides <- filter(anno, neuropeptide == 1)$gene
neuropeptides <- c(neuropeptides, "Slc32a1", "Slc17a6", "Slc16a1", "Slc6a3")

receptors <- filter(anno, receptor == 1)$gene

df <- left_join(enrichment, genes, by = "gene_id")
df <- left_join(df, mean_cpm, by = c("gene_id", "gene_name", "gene_biotype"))
df <- filter(df, gene_name %in% neuropeptides)

```


```{r, fig.width = 1.2, fig.height = 2.5}
# use Life Aquatic color palette per Paul's request
color_palette <- wesanderson::wes_palette("Zissou1", 100, type = "continuous")
# reverse palette to make blue high and red low
color_palette <- color_palette[length(color_palette):1]

df %>%
  filter(padj < 0.05 | gene_name %in% c("Slc17a6", "Slc32a1", "Chat",
                                        "Slc16a1", "Slc6a3")) %>%
  arrange(desc(log2FoldChange)) %>%
  mutate(gene_name = factor(gene_name, levels = gene_name)) %>%
  ggplot(aes(x = "", y = fct_rev(gene_name), color = log2FoldChange)) +
  geom_point(aes(size = log2(Bead))) +
  scale_color_gradientn(colors = color_palette,
                        name = expression("log"[2]*"FC")) +
  scale_radius(name = expression("log"[2]*"CPM")) +
  theme_void() +
  theme(axis.text.y = element_text(hjust = 1, vjust = 0.5,
                                   size = 8),
        legend.key.width = unit(0.02, "in"),
        legend.key.height = unit(0.1, "in"),
        legend.text = element_text(size = 5),
        legend.title = element_text(size = 7))

ggsave("figures/panels/heatmap.png", width = 1.2, height = 2.5,
       units = "in", dpi = 400)
```


\pagebreak

## GDF15 effects
Now using DESeq2 to identify the effects of GDF15.
```{r regulation}
bead_info <- filter(info, cells == "Bead")
mtx2 <- mtx[, bead_info$Sample_ID]

print("Setting up DESeq2 object ...")
dds_gdf15 <- DESeq2::DESeqDataSetFromMatrix(mtx2, bead_info, ~ treatment)
print("Running DESeq ...")
dds_gdf15 <- DESeq2::DESeq(dds_gdf15)

print("Finding regulated genes ...")
regulation <- DESeq2::results(dds_gdf15) %>%
  as.data.frame() %>%
  rownames_to_column("gene_id")

regulated <- filter(regulation, padj < 0.05)$gene_id
```

`r length(regulated)` genes are significant regulated by GDF15 treatment in the Bead.

```{r}
cpm_tx <- map(unique(bead_info$treatment),
    ~ rowMeans(log2(cpm[,filter(bead_info, treatment == .x)$Sample_ID]+1))) %>%
  as.data.frame() %>%
  set_names(unique(bead_info$treatment)) %>%
  mutate("gene_id" = cpm$gene_id)
```

```{r, fig.width = 3, fig.height = 3}
map(unique(bead_info$treatment),
    ~ rowMeans(log2(cpm[,filter(bead_info, treatment == .x)$Sample_ID]+1))) %>%
  as.data.frame() %>%
  set_names(unique(bead_info$treatment)) %>%
  mutate_all(~2^.x - 1) %>%
  mutate("gene_id" = cpm$gene_id) %>%
  filter(gene_id %in% filter(regulation, !is.na(padj))$gene_id) %>%
  left_join(., regulation, by = "gene_id") %>%
  left_join(., genes, by = "gene_id") %>%
  ggplot(aes(x = Vehicle, y = GDF15, color = gene_id %in% regulated)) +
  geom_abline(aes(intercept = 0, slope = 1)) +
  geom_point(alpha = 0.4, stroke = 0, show.legend = FALSE) +
  scale_color_manual(values = c("gray30", "red4")) +
  scale_x_continuous(trans = "log2", limits = c(1, NA), expand = c(0, 0)) +
  scale_y_continuous(trans = "log2", limits = c(1, NA), expand = c(0, 0)) +
  theme_bw() +
  theme(panel.grid = element_blank(),
        axis.text = element_text(color = "black")) +
  xlab("Vehicle Bead expression (CPM)") +
  ylab("GDF15 Bead expression (CPM)")
```

```{r}
left_join(regulation, genes, by = "gene_id") %>%
  left_join(cpm_tx, by = "gene_id") %>%
  select(gene_name, Vehicle, GDF15, log2FoldChange, padj) %>%
  arrange(padj) %>%
  filter(padj < 0.99) %>%
  slice(1:10) %>%
  mutate_at(vars(Vehicle, GDF15, log2FoldChange), ~ round(.x, 2)) %>%
  mutate(padj = format(padj, scientific = TRUE, digits = 2)) %>%
  rename("Gene" = gene_name, "Regulation (log2)" = log2FoldChange, "P" = padj) %>%
  kable(caption = "Top regulated genes in Bead") %>%
  kable_styling(latex_options = "hold_position")
```


```{r, fig.width = 2.5, fig.height = 2}
left_join(cpm, genes, by = "gene_id") %>%
  filter(gene_name %in% c("Pdyn", "Penk")) %>%
  select(-gene_id, -gene_biotype) %>%
  pivot_longer(-gene_name, names_to = "Sample_ID", values_to = "CPM") %>%
  left_join(info, by = "Sample_ID") %>%
  filter(cells == "Bead") %>%
  ggplot(aes(x = treatment, y = CPM)) +
  geom_boxplot() +
  geom_jitter() +
  xlab(NULL) +
  ylab("Expression (CPM)") +
  scale_y_continuous(trans = "log2") +
  theme_classic() +
  facet_wrap(~gene_name, scales = "free_y") +
  theme(strip.background = element_blank(),
        strip.text = element_text(size = 10),
        axis.text.x = element_text(size = 10, color = "black",
                                   hjust = 1, vjust = 0.5, angle = 90),
        axis.text.y = element_text(size = 8, color = "black"))
```

\pagebreak

# Files
I'm saving the `R` image as `analysis.Rdata` for future easy loading.

```{r}
save.image("analysis.Rdata", compress = TRUE)
```

\pagebreak

# Session info
```{r}
sessionInfo()
```
